tests/templates/test_subroutines/test_select_pauli_rot.py

```diff
@@ -0,0 +1,292 @@
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Unit tests for the SelectPauliRot template.
+"""
+import numpy as np
+
+# pylint: disable=too-many-arguments,too-few-public-methods
+import pytest
+
+import pennylane as qml
+from pennylane import numpy as np
+
+
+class TestSelectPauliRot:
+
+    def test_standard_validity(self):
+        """Check the operation using the assert_valid function."""
+
+        wires = qml.registers({"control_wires": 3, "target_wire": 1})
+
+        op = qml.SelectPauliRot(
+            angles=qml.math.ones(8),
+            control_wires=wires["control_wires"],
+            target_wire=wires["target_wire"],
+            rot_axis="X",
+        )
+
+        qml.ops.functions.assert_valid(op)
+
+    @pytest.mark.parametrize(
+        ("angles", "rot_axis", "target_wire", "msg_match"),
+        [
+            (
+                np.array([1.0, 2.0, 3.0, 4.0]),
+                "K",
+                2,
+                "'rot_axis' can only take the values 'X', 'Y' and 'Z'.",
+            ),
+            (
+                np.array([1.0]),
+                "Z",
+                2,
+                "Number of angles must",
+            ),
+            (
+                np.array([1.0, 2.0, 3.0, 4.0]),
+                "Z",
+                [2, 3],
+                "Only one target wire can",
+            ),
+        ],
+    )
+    def test_SelectPauliRot_error(self, angles, rot_axis, target_wire, msg_match):
+        """Test that proper errors are raised for SelectPauliRot"""
+
+        wires = qml.registers({"control_wires": 2})
+
+        with pytest.raises(ValueError, match=msg_match):
+            qml.SelectPauliRot(
+                angles=angles,
+                control_wires=wires["control_wires"],
+                target_wire=target_wire,
+                rot_axis=rot_axis,
+            )
+
+    @pytest.mark.parametrize(
+        ("angles", "rot_axis"),
+        [
+            (np.array([1, 2, 3, 4, 5, 6, 7, 8]), "Z"),
+            (np.array([12, 2, 3, 4, 5.4, 6, 7, 8]), "X"),
+            (np.array([1, 2, -3, 4, 5, 26, 7, 8]), "Y"),
+        ],
+    )
+    def test_correctness(self, angles, rot_axis):
+        """Tests the correctness of the SelectPauliRot template.
+        This is done by comparing the results with the naive Select(Rotation) decomposition
+        """
+
+        dev = qml.device("default.qubit", wires=4)
+
+        gate = {"Z": qml.RZ, "Y": qml.RY, "X": qml.RX}
+
+        # Check that applying the SelectPauliRot and adjoint(Select) to an state,
+        # does not modify the state
+        qs = qml.tape.QuantumScript(
+            [
+                qml.Hadamard(0),
+                qml.Hadamard(1),
```

Comment:
Why don't we apply Hadamard on the 3rd wire?

Reply:
I'm just generating an arbitrary state, I could add it too or use any other gates